File: src/main/java/org/organicdesign/fp/collections/AbstractUnmodIterable.java
package org.organicdesign.fp.collections;

import org.jetbrains.annotations.NotNull;

/**
 Implements equals and hashCode() methods compatible with all java.util collections (this
 algorithm is not order-dependent) and toString which takes the name of the sub-class.
 */
public abstract class AbstractUnmodIterable<T> implements UnmodIterable<T> {
    @Override public int hashCode() { return UnmodIterable.hash(this); }
    @Override public @NotNull String toString() {
        return UnmodIterable.toString(getClass().getSimpleName(), this);
    }
}

File: src/main/java/org/organicdesign/fp/collections/AbstractUnmodIterable.kt
package org.organicdesign.fp.collections

/**
 * Implements equals and hashCode() methods compatible with all java.util collections (this
 * algorithm is not order-dependent) and toString which takes the name of the sub-class.
 * You must override equals!
 */
abstract class AbstractUnmodIterable<T> : UnmodIterable<T> {
    abstract override fun equals(other: Any?): Boolean

    override fun hashCode(): Int = UnmodIterable.hash(this)

    override fun toString(): String = UnmodIterable.toString(javaClass.simpleName, this)
}
